dwave/optimization/model.pyi

```diff
@@ -12,29 +12,166 @@
+        lower_bound: typing.Optional[int] = None,
+        upper_bound: typing.Optional[int] = None,
+        ) -> IntegerVariable: ...
+
+    def into_file(
+        self,
+        file: typing.Union[typing.BinaryIO, typing.ByteString, str],
+        *,
+        max_num_states: int = 0,
+        only_decision: bool = False,
+        ): ...
+
+    def is_locked(self) -> bool: ...
+    def iter_constraints(self) -> typing.Iterator[ArraySymbol]: ...
+    def iter_decisions(self) -> typing.Iterator[Symbol]: ...
+    def iter_symbols(self) -> typing.Iterator[Symbol]: ...
     def list(self, n: int) -> ListVariable: ...
-    def lock(self) -> None: ...
+    def lock(self) -> typing.ContextManager: ...
+    def minimize(self, value: ArraySymbol): ...
+    def num_constraints(self) -> int: ...
+    def num_decisions(self) -> int: ...
     def num_nodes(self) -> int: ...
-    def resize_states(self, n: int): ...
+    def num_symbols(self) -> int: ...
+
+    # dev note: this is underspecified, but it would be quite complex to fully
+    # specify the linear/quadratic so let's leave it alone for now.
+    def quadratic_model(self, x: ArraySymbol, quadratic, linear=None) -> QuadraticModel: ...
 
+    def set(
+        self,
+        n: int,
+        min_size: int = 0,
+        max_size: typing.Optional[int] = None,
+        ) -> SetVariable: ...
 
-class Constant(ArraySymbol):
-    ...
+    def state_size(self) -> int: ...
 
+    def to_file(
+        self,
+        *,
+        max_num_states: int = 0,
+        only_decision: bool = False,
+        ) -> typing.BinaryIO: ...
 
-class ListVariable(ArraySymbol):
-    def set_state(self, index: int, state: numpy.typing.ArrayLike) -> None: ...
+    # networkx might not be installed, so we just say we return an object.
+    def to_networkx(self) -> object: ...
 
+    def unlock(self): ...
 
-class StatesView:
+
+class States:
+    def __init__(self, model: Model): ...
     def __len__(self) -> int: ...
-    def __getitem__(self, index: int) -> numpy.typing.NDArray[numpy.float64]: ...
+    def clear(self): ...
+
+    def from_file(
+        self,
+        file: typing.Union[typing.BinaryIO, typing.ByteString, str],
+        *,
+        replace: bool = True,
+        check_header: bool = True,
+        ) -> Model: ...
+
+    def from_future(self, future: object, result_hook: typing.Callable): ...
+    def initialize(self): ...
+
+    def into_file(
+        self,
+        file: typing.Union[typing.BinaryIO, typing.ByteString, str],
+        ): ...
+
+    def resize(self, n: int): ...
+    def resolve(self): ...
+    def size(self) -> int: ...
+    def to_file(self) -> typing.BinaryIO: ...
+
+
+class Symbol:
+    def __init__(self, *args, **kwargs) -> typing.NoReturn: ...
+    def equals(self, other: Symbol) -> bool: ...
+    def expired(self) -> bool: ...
+    def has_state(self, index: int = 0) -> bool: ...
+    def iter_predecessors(self) -> typing.Iterator[Symbol]: ...
```

Comment:
Why is this an iterator? The N successors have a single predecessor

Reply:
it's always `for pred in symbol.iter_predecessors(): ...`. I.e. an iterator.